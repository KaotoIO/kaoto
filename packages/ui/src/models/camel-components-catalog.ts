import { CatalogKind } from './catalog-kind';

export interface ICamelComponentDefinition {
  component: ICamelComponent;
  componentProperties: ICamelComponentProperties;
  headers: ICamelComponentHeaders;
  properties: ICamelComponentProperties;
}

export interface ICamelComponent {
  kind: CatalogKind.Component;
  name: string;
  title: string;
  description: string;
  deprecated: boolean;
  firstVersion?: string;
  label: string;
  javaType?: string;
  supportLevel?: string;
  groupId?: string;
  artifactId?: string;
  version: string;
  scheme?: string;
  extendsScheme?: string;
  syntax?: string;
  async?: boolean;
  api?: boolean;
  consumerOnly?: boolean;
  producerOnly?: boolean;
  lenientProperties?: boolean;
}

export interface CatalogItemCommon {
  index: number;
  kind: string;
  displayName: string;
  group: string;
  label: string;
  required: boolean;
  javaType: string;
  deprecated: boolean;
  deprecationNote: string;
  autowired: boolean;
  secret: boolean;
  description: string;
}

export interface ICamelComponentProperties {
  type: string;
}

export interface ICamelComponentHeaders extends CatalogItemCommon {
  constantName: string;
}
