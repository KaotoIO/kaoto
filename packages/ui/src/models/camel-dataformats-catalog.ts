import { ICamelProcessorModel, ICamelProcessorProperty } from './camel-processors-catalog';
import { KaotoSchemaDefinition } from './kaoto-schema';

export interface ICamelDataformatDefinition {
  model: ICamelDataformatModel;
  properties: Record<string, ICamelDataformatProperty>;
  propertiesSchema: KaotoSchemaDefinition['schema'];
}

export type ICamelDataformatModel = ICamelProcessorModel;

export type ICamelDataformatProperty = ICamelProcessorProperty;
