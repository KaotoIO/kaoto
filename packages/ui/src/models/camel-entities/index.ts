export * from './base-entity';

/** Camel Route definitions */
export * from './flows/route';
