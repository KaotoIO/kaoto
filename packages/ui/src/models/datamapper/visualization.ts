import { IDocument, IField, PrimitiveDocument } from './document';
import { ExpressionItem, FieldItem, IFunctionDefinition, MappingItem, MappingParentType, MappingTree } from './mapping';
import { DocumentType, NodePath } from './path';
import { RefObject } from 'react';

export interface NodeData {
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export interface TargetNodeData extends NodeData {
  mappingTree: MappingTree;
  mapping?: MappingParentType;
}

export type SourceNodeDataType = DocumentNodeData | FieldNodeData;
export type TargetNodeDataType = TargetDocumentNodeData | TargetFieldNodeData;

export class DocumentNodeData implements NodeData {
  constructor(document: IDocument) {
    this.title = document.documentId;
    this.id = `doc-${document.documentType}-${document.documentId}`;
    this.path = NodePath.fromDocument(document.documentType, document.documentId);
    this.document = document;
    this.isSource = document.documentType !== DocumentType.TARGET_BODY;
    this.isPrimitive = document instanceof PrimitiveDocument;
  }

  document: IDocument;
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetDocumentNodeData extends DocumentNodeData implements TargetNodeData {
  constructor(document: IDocument, mappingTree: MappingTree) {
    super(document);
    this.mappingTree = mappingTree;
    this.mapping = mappingTree;
  }
  mappingTree: MappingTree;
  mapping: MappingTree;
}

export class FieldNodeData implements NodeData {
  constructor(
    public parent: NodeData,
    public field: IField,
  ) {
    this.title = field.name;
    this.id = field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isSource = parent.isSource;
    this.isPrimitive = parent.isPrimitive;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetFieldNodeData extends FieldNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
    public mapping?: FieldItem,
  ) {
    super(parent, field);
    this.mappingTree = parent.mappingTree;
  }
  mappingTree: MappingTree;
}

export class MappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: MappingItem,
  ) {
    this.title = mapping.name;
    this.id = mapping.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource = false;
  isPrimitive: boolean;
  mappingTree: MappingTree;
}

export class FieldItemNodeData extends MappingNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: FieldItem,
  ) {
    super(parent, mapping);
    this.title = mapping.field.name;
    this.field = mapping.field;
  }
  public field: IField;
}

export class AddMappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
  ) {
    const ID_PREFIX = 'add-mapping-';
    this.id = ID_PREFIX + field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.title = field.name;
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }
  id: string;
  isPrimitive: boolean;
  isSource = false;
  mappingTree: MappingTree;
  path: NodePath;
  title: string;
}

class SimpleNodePath extends NodePath {
  constructor(public path: string) {
    super();
  }
  toString() {
    return this.path;
  }
}
export class EditorNodeData implements NodeData {
  constructor(public mapping: ExpressionItem) {}
  id: string = 'editor';
  isPrimitive: boolean = false;
  isSource: boolean = false;
  path: NodePath = new SimpleNodePath('Editor');
  title: string = 'Editor';
}

export class FunctionNodeData implements NodeData {
  constructor(public functionDefinition: IFunctionDefinition) {
    this.id = functionDefinition.name;
    this.title = functionDefinition.displayName;
    this.path = new SimpleNodePath('Func:' + functionDefinition.name);
  }

  id: string;
  isPrimitive: boolean = false;
  isSource: boolean = true;
  path: NodePath;
  title: string;
}

export interface IMappingLink {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected: boolean;
}

export interface NodeReference {
  path: string;
  isSource: boolean;
  headerRef: HTMLDivElement | null;
  containerRef: HTMLDivElement | null;
}

export type LineCoord = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

export type LineProps = LineCoord & {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected?: boolean;
  svgRef?: RefObject<SVGSVGElement>;
};
