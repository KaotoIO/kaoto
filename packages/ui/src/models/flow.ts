export interface FlowModel {
  id: string;
  type: string;
  model: unknown;
}
