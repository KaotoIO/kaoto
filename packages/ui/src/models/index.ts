export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
