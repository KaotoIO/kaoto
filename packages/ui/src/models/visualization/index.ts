export * from './base-visual-entity';
export * from './visualization-node';
export * from './flows';
