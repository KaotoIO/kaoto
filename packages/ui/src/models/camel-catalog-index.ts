import { ICamelComponentDefinition } from './camel-components-catalog';
import { ICamelProcessorDefinition } from './camel-processors-catalog';
import { CatalogKind } from './catalog-kind';
import { IKameletDefinition } from './kamelets-catalog';

export interface CamelCatalogIndex {
  catalogs: Catalogs;
  schemas: CatalogEntry[];
}

export interface Catalogs {
  models: CatalogEntry;
  components: CatalogEntry;
  languages: CatalogEntry;
  dataformats: CatalogEntry;
  kamelets: CatalogEntry;
}

export interface CatalogEntry {
  name: string;
  version: string;
  file: string;
}

export type CatalogTypes = Record<string, ICamelComponentDefinition | ICamelProcessorDefinition | IKameletDefinition>;

export interface ComponentsCatalog {
  [CatalogKind.Component]?: Record<string, ICamelComponentDefinition>;
  [CatalogKind.Processor]?: Record<string, ICamelProcessorDefinition>;
  [CatalogKind.Kamelet]?: Record<string, IKameletDefinition>;
  [key: string]: unknown;
}

export interface CatalogCamelComponent {
  type: CatalogKind.Component;
  definition: ICamelComponentDefinition;
}

export interface CatalogCamelProcessor {
  type: CatalogKind.Processor;
  definition: ICamelProcessorDefinition;
}

export interface CatalogKamelet {
  type: CatalogKind.Kamelet;
  definition: IKameletDefinition;
}
