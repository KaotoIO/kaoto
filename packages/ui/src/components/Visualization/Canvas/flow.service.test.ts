import { CamelRouteVisualEntity, createVisualizationNode } from '../../../models/visualization';
import { FlowService } from './flow.service';

describe('FlowService', () => {
  beforeEach(() => {
    FlowService.nodes = [];
    FlowService.edges = [];
  });

  it('should start with an empty nodes array', () => {
    expect(FlowService.nodes).toEqual([]);
  });

  it('should start with an empty edges array', () => {
    expect(FlowService.edges).toEqual([]);
  });

  describe('getFlowDiagram', () => {
    it('should return nodes and edges for a simple VisualizationNode', () => {
      const vizNode = createVisualizationNode('node', {});

      const { nodes, edges } = FlowService.getFlowDiagram('test', vizNode);

      expect(nodes).toMatchSnapshot();
      expect(edges).toMatchSnapshot();
    });

    it('should return nodes and edges for a group with children', () => {
      const groupVizNode = createVisualizationNode('group', { isGroup: true });
      const child1VizNode = createVisualizationNode('child1', {});
      const child2VizNode = createVisualizationNode('child2', {});
      groupVizNode.addChild(child1VizNode);
      groupVizNode.addChild(child2VizNode);

      const { nodes, edges } = FlowService.getFlowDiagram('test', groupVizNode);

      expect(nodes).toMatchSnapshot();
      expect(edges).toMatchSnapshot();
    });

    it('should return nodes and edges for a two-nodes VisualizationNode', () => {
      const vizNode = createVisualizationNode('node', {});
      const childNode = createVisualizationNode('child', {});
      vizNode.addChild(childNode);

      const { nodes, edges } = FlowService.getFlowDiagram('test', vizNode);

      expect(nodes).toMatchSnapshot();
      expect(edges).toMatchSnapshot();
    });

    it('should return nodes and edges for a multiple nodes VisualizationNode', () => {
      const vizNode = createVisualizationNode('node', {});

      const setHeaderNode = createVisualizationNode('set-header', {});
      vizNode.setNextNode(setHeaderNode);
      setHeaderNode.setPreviousNode(vizNode);

      const choiceNode = createVisualizationNode('choice', {});
      setHeaderNode.setNextNode(choiceNode);
      choiceNode.setPreviousNode(setHeaderNode);

      const directNode = createVisualizationNode('direct', {});
      choiceNode.setNextNode(directNode);
      directNode.setPreviousNode(choiceNode);

      const whenNode = createVisualizationNode('when', {});
      choiceNode.addChild(whenNode);

      const otherwiseNode = createVisualizationNode('otherwise', {});
      choiceNode.addChild(otherwiseNode);

      const whenLeafNode = createVisualizationNode('when-leaf', {});
      whenNode.addChild(whenLeafNode);

      const processNode = createVisualizationNode('process', {});
      otherwiseNode.addChild(processNode);
      const logNode = createVisualizationNode('log', {});
      processNode.addChild(logNode);

      const { nodes, edges } = FlowService.getFlowDiagram('test', vizNode);

      expect(nodes).toMatchSnapshot();
      expect(edges).toMatchSnapshot();
    });

    it('should return a group node for a multiple nodes VisualizationNode with a group', () => {
      const routeNode = new CamelRouteVisualEntity({ from: { uri: 'timer:clock', steps: [] } }).toVizNode();

      const { nodes, edges } = FlowService.getFlowDiagram('test', routeNode);

      expect(nodes).toHaveLength(3);
      expect(nodes[0].data?.vizNode?.data.path).toEqual('route.from');
      expect(nodes[1].data?.vizNode?.data.path).toEqual('route.from.steps.0.placeholder');
      expect(nodes[2].data?.vizNode?.data.path).toEqual('route');

      expect(edges).toHaveLength(1);
      expect(edges[0].source).toEqual('test|route.from');
      expect(edges[0].target).toEqual('test|route.from.steps.0.placeholder');

      const group = nodes[nodes.length - 1];
      expect(group.children).toEqual(['test|route.from', 'test|route.from.steps.0.placeholder']);
      expect(group.group).toBeTruthy();
    });
  });
});
