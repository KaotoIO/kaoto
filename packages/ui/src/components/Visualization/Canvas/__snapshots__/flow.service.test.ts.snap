// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowService getFlowDiagram should return nodes and edges for a group with children 1`] = `
[
  {
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": undefined,
        "data": {},
        "id": "child1-1234",
        "nextNode": undefined,
        "parentNode": VisualizationNode {
          "DISABLED_NODE_INTERACTION": {
            "canBeDisabled": false,
            "canHaveChildren": false,
            "canHaveNextStep": false,
            "canHavePreviousStep": false,
            "canHaveSpecialChildren": false,
            "canRemoveFlow": false,
            "canRemoveStep": false,
            "canReplaceStep": false,
          },
          "children": [
            [Circular],
            VisualizationNode {
              "DISABLED_NODE_INTERACTION": {
                "canBeDisabled": false,
                "canHaveChildren": false,
                "canHaveNextStep": false,
                "canHavePreviousStep": false,
                "canHaveSpecialChildren": false,
                "canRemoveFlow": false,
                "canRemoveStep": false,
                "canReplaceStep": false,
              },
              "children": undefined,
              "data": {},
              "id": "child2-1234",
              "nextNode": undefined,
              "parentNode": [Circular],
              "previousNode": undefined,
              "title": "child2",
            },
          ],
          "data": {
            "isGroup": true,
          },
          "id": "group-1234",
          "nextNode": undefined,
          "parentNode": undefined,
          "previousNode": undefined,
          "title": "group",
        },
        "previousNode": undefined,
        "title": "child1",
      },
    },
    "height": 75,
    "id": "child1-1234",
    "parentNode": "group-1234",
    "shape": "rect",
    "type": "node",
    "width": 90,
  },
  {
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": undefined,
        "data": {},
        "id": "child2-1234",
        "nextNode": undefined,
        "parentNode": VisualizationNode {
          "DISABLED_NODE_INTERACTION": {
            "canBeDisabled": false,
            "canHaveChildren": false,
            "canHaveNextStep": false,
            "canHavePreviousStep": false,
            "canHaveSpecialChildren": false,
            "canRemoveFlow": false,
            "canRemoveStep": false,
            "canReplaceStep": false,
          },
          "children": [
            VisualizationNode {
              "DISABLED_NODE_INTERACTION": {
                "canBeDisabled": false,
                "canHaveChildren": false,
                "canHaveNextStep": false,
                "canHavePreviousStep": false,
                "canHaveSpecialChildren": false,
                "canRemoveFlow": false,
                "canRemoveStep": false,
                "canReplaceStep": false,
              },
              "children": undefined,
              "data": {},
              "id": "child1-1234",
              "nextNode": undefined,
              "parentNode": [Circular],
              "previousNode": undefined,
              "title": "child1",
            },
            [Circular],
          ],
          "data": {
            "isGroup": true,
          },
          "id": "group-1234",
          "nextNode": undefined,
          "parentNode": undefined,
          "previousNode": undefined,
          "title": "group",
        },
        "previousNode": undefined,
        "title": "child2",
      },
    },
    "height": 75,
    "id": "child2-1234",
    "parentNode": "group-1234",
    "shape": "rect",
    "type": "node",
    "width": 90,
  },
  {
    "children": [
      "child1-1234",
      "child2-1234",
    ],
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": [
          VisualizationNode {
            "DISABLED_NODE_INTERACTION": {
              "canBeDisabled": false,
              "canHaveChildren": false,
              "canHaveNextStep": false,
              "canHavePreviousStep": false,
              "canHaveSpecialChildren": false,
              "canRemoveFlow": false,
              "canRemoveStep": false,
              "canReplaceStep": false,
            },
            "children": undefined,
            "data": {},
            "id": "child1-1234",
            "nextNode": undefined,
            "parentNode": [Circular],
            "previousNode": undefined,
            "title": "child1",
          },
          VisualizationNode {
            "DISABLED_NODE_INTERACTION": {
              "canBeDisabled": false,
              "canHaveChildren": false,
              "canHaveNextStep": false,
              "canHavePreviousStep": false,
              "canHaveSpecialChildren": false,
              "canRemoveFlow": false,
              "canRemoveStep": false,
              "canReplaceStep": false,
            },
            "children": undefined,
            "data": {},
            "id": "child2-1234",
            "nextNode": undefined,
            "parentNode": [Circular],
            "previousNode": undefined,
            "title": "child2",
          },
        ],
        "data": {
          "isGroup": true,
        },
        "id": "group-1234",
        "nextNode": undefined,
        "parentNode": undefined,
        "previousNode": undefined,
        "title": "group",
      },
    },
    "group": true,
    "id": "group-1234",
    "label": "group-1234",
    "parentNode": undefined,
    "style": {
      "padding": 40,
    },
    "type": "group",
  },
]
`;

exports[`FlowService getFlowDiagram should return nodes and edges for a group with children 2`] = `[]`;

exports[`FlowService getFlowDiagram should return nodes and edges for a multiple nodes VisualizationNode 1`] = `
[
  {
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": undefined,
        "data": {},
        "id": "node-1234",
        "nextNode": VisualizationNode {
          "DISABLED_NODE_INTERACTION": {
            "canBeDisabled": false,
            "canHaveChildren": false,
            "canHaveNextStep": false,
            "canHavePreviousStep": false,
            "canHaveSpecialChildren": false,
            "canRemoveFlow": false,
            "canRemoveStep": false,
            "canReplaceStep": false,
          },
          "children": undefined,
          "data": {},
          "id": "set-header-1234",
          "nextNode": VisualizationNode {
            "DISABLED_NODE_INTERACTION": {
              "canBeDisabled": false,
              "canHaveChildren": false,
              "canHaveNextStep": false,
              "canHavePreviousStep": false,
              "canHaveSpecialChildren": false,
              "canRemoveFlow": false,
              "canRemoveStep": false,
              "canReplaceStep": false,
            },
            "children": [
              VisualizationNode {
                "DISABLED_NODE_INTERACTION": {
                  "canBeDisabled": false,
                  "canHaveChildren": false,
                  "canHaveNextStep": false,
                  "canHavePreviousStep": false,
                  "canHaveSpecialChildren": false,
                  "canRemoveFlow": false,
                  "canRemoveStep": false,
                  "canReplaceStep": false,
                },
                "children": [
                  VisualizationNode {
                    "DISABLED_NODE_INTERACTION": {
                      "canBeDisabled": false,
                      "canHaveChildren": false,
                      "canHaveNextStep": false,
                      "canHavePreviousStep": false,
                      "canHaveSpecialChildren": false,
                      "canRemoveFlow": false,
                      "canRemoveStep": false,
                      "canReplaceStep": false,
                    },
                    "children": undefined,
                    "data": {},
                    "id": "when-leaf-1234",
                    "nextNode": undefined,
                    "parentNode": [Circular],
                    "previousNode": undefined,
                    "title": "when-leaf",
                  },
                ],
                "data": {},
                "id": "when-1234",
                "nextNode": undefined,
                "parentNode": [Circular],
                "previousNode": undefined,
                "title": "when",
              },
              VisualizationNode {
                "DISABLED_NODE_INTERACTION": {
                  "canBeDisabled": false,
                  "canHaveChildren": false,
                  "canHaveNextStep": false,
                  "canHavePreviousStep": false,
                  "canHaveSpecialChildren": false,
                  "canRemoveFlow": false,
                  "canRemoveStep": false,
                  "canReplaceStep": false,
                },
                "children": [
                  VisualizationNode {
                    "DISABLED_NODE_INTERACTION": {
                      "canBeDisabled": false,
                      "canHaveChildren": false,
                      "canHaveNextStep": false,
                      "canHavePreviousStep": false,
                      "canHaveSpecialChildren": false,
                      "canRemoveFlow": false,
                      "canRemoveStep": false,
                      "canReplaceStep": false,
                    },
                    "children": [
                      VisualizationNode {
                        "DISABLED_NODE_INTERACTION": {
                          "canBeDisabled": false,
                          "canHaveChildren": false,
                          "canHaveNextStep": false,
                          "canHavePreviousStep": false,
                          "canHaveSpecialChildren": false,
                          "canRemoveFlow": false,
                          "canRemoveStep": false,
                          "canReplaceStep": false,
                        },
                        "children": undefined,
                        "data": {},
                        "id": "log-1234",
                        "nextNode": undefined,
                        "parentNode": [Circular],
                        "previousNode": undefined,
                        "title": "log",
                      },
                    ],
                    "data": {},
                    "id": "process-1234",
                    "nextNode": undefined,
                    "parentNode": [Circular],
                    "previousNode": undefined,
                    "title": "process",
                  },
                ],
                "data": {},
                "id": "otherwise-1234",
                "nextNode": undefined,
                "parentNode": [Circular],
                "previousNode": undefined,
                "title": "otherwise",
              },
            ],
            "data": {},
            "id": "choice-1234",
            "nextNode": VisualizationNode {
              "DISABLED_NODE_INTERACTION": {
                "canBeDisabled": false,
                "canHaveChildren": false,
                "canHaveNextStep": false,
                "canHavePreviousStep": false,
                "canHaveSpecialChildren": false,
                "canRemoveFlow": false,
                "canRemoveStep": false,
                "canReplaceStep": false,
              },
              "children": undefined,
              "data": {},
              "id": "direct-1234",
              "nextNode": undefined,
              "parentNode": undefined,
              "previousNode": [Circular],
              "title": "direct",
            },
            "parentNode": undefined,
            "previousNode": [Circular],
            "title": "choice",
          },
          "parentNode": undefined,
          "previousNode": [Circular],
          "title": "set-header",
        },
        "parentNode": undefined,
        "previousNode": undefined,
        "title": "node",
      },
    },
    "height": 75,
    "id": "node-1234",
    "parentNode": undefined,
    "shape": "rect",
    "type": "node",
    "width": 90,
  },
]
`;

exports[`FlowService getFlowDiagram should return nodes and edges for a multiple nodes VisualizationNode 2`] = `
[
  {
    "edgeStyle": "solid",
    "id": "node-1234-to-set-header-1234",
    "source": "node-1234",
    "target": "set-header-1234",
    "type": "edge",
  },
]
`;

exports[`FlowService getFlowDiagram should return nodes and edges for a simple VisualizationNode 1`] = `
[
  {
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": undefined,
        "data": {},
        "id": "node-1234",
        "nextNode": undefined,
        "parentNode": undefined,
        "previousNode": undefined,
        "title": "node",
      },
    },
    "height": 75,
    "id": "node-1234",
    "parentNode": undefined,
    "shape": "rect",
    "type": "node",
    "width": 90,
  },
]
`;

exports[`FlowService getFlowDiagram should return nodes and edges for a simple VisualizationNode 2`] = `[]`;

exports[`FlowService getFlowDiagram should return nodes and edges for a two-nodes VisualizationNode 1`] = `
[
  {
    "data": {
      "vizNode": VisualizationNode {
        "DISABLED_NODE_INTERACTION": {
          "canBeDisabled": false,
          "canHaveChildren": false,
          "canHaveNextStep": false,
          "canHavePreviousStep": false,
          "canHaveSpecialChildren": false,
          "canRemoveFlow": false,
          "canRemoveStep": false,
          "canReplaceStep": false,
        },
        "children": [
          VisualizationNode {
            "DISABLED_NODE_INTERACTION": {
              "canBeDisabled": false,
              "canHaveChildren": false,
              "canHaveNextStep": false,
              "canHavePreviousStep": false,
              "canHaveSpecialChildren": false,
              "canRemoveFlow": false,
              "canRemoveStep": false,
              "canReplaceStep": false,
            },
            "children": undefined,
            "data": {},
            "id": "child-1234",
            "nextNode": undefined,
            "parentNode": [Circular],
            "previousNode": undefined,
            "title": "child",
          },
        ],
        "data": {},
        "id": "node-1234",
        "nextNode": undefined,
        "parentNode": undefined,
        "previousNode": undefined,
        "title": "node",
      },
    },
    "height": 75,
    "id": "node-1234",
    "parentNode": undefined,
    "shape": "rect",
    "type": "node",
    "width": 90,
  },
]
`;

exports[`FlowService getFlowDiagram should return nodes and edges for a two-nodes VisualizationNode 2`] = `[]`;
